shared/src/lib.rs

```diff
@@ -116,3 +117,29 @@ pub fn get_local_addrs() -> Result<impl Iterator<Item = std::net::IpAddr>, io::E
         })
         .take(10))
 }
+
+pub trait IpNetExt {
+    fn is_assignable(&self, ip: &IpAddr) -> bool;
+
+    #[allow(clippy::new_ret_no_self)]
+    fn new(ip: IpAddr, prefix_len: u8) -> Result<IpNet, PrefixLenError>;
+}
+
+impl IpNetExt for IpNet {
+    fn new(ip: IpAddr, prefix_len: u8) -> Result<IpNet, PrefixLenError> {
+        Ok(match ip {
+            IpAddr::V4(a) => Ipv4Net::new(a, prefix_len)?.into(),
+            IpAddr::V6(a) => Ipv6Net::new(a, prefix_len)?.into(),
+        })
+    }
+
+    fn is_assignable(&self, ip: &IpAddr) -> bool {
+        self.contains(ip)
+            && match self {
+                IpNet::V4(_) => {
+                    self.prefix_len() >= 31 || (ip != &self.network() && ip != &self.broadcast())
+                },
+                IpNet::V6(_) => self.prefix_len() >= 127 || ip != &self.network(),
+            }
```

Comment:
Curiosity question: why do we need to make exception for the "full prefix" and "full minus one prefix"? Without this exception, such CIDRs would not be assignable at all in IPv4 case and very limited in IPv6 case - perhaps that's the reason?

Reply:
Yep, exactly. It's more likely an issue in the IPv4 space, but in a /31 the only two addresses in that CIDR would technically be the network and broadcast address, but there is an RFC that clarifies that /31 is special-cased in the same way that /32 is for that reason. I'll have to find the specific RFC that states this.

Reply:
Oh I see, this is special-cased in the standard. Thanks!